src/aws-cpp-sdk-core/source/utils/stream/ConcurrentStreamBuf.cpp

```diff
@@ -67,19 +93,32 @@ namespace Aws
 
             int ConcurrentStreamBuf::underflow()
             {
+                bool closeStream = false;
                 {
                     std::unique_lock<std::mutex> lock(m_lock);
-                    m_signal.wait(lock, [this]{ return m_backbuf.empty() == false || m_eof; });
 
-                    if (m_eof && m_backbuf.empty())
+                    if (!m_eofInput)
                     {
-                        return std::char_traits<char>::eof();
+                        m_signal.wait_for(lock, std::chrono::milliseconds(1000), [this] { return m_backbuf.empty() == false || m_eofInput; });
```

Comment:
this is technically non-compliant with std::
underflow must fill get buffer or return EOF, 

but we may have nothing to fill and blocking will make things worse;
EOF will terminate the HTTP2 conversation too.